ring/replication_set.go

```diff
@@ -300,6 +319,13 @@ func DoUntilQuorumWithoutSuccessfulContextCancellation[T any](ctx context.Contex
 			resultTracker.startAdditionalRequests()
 		case result := <-resultsChan:
 			resultsRemaining--
+
+			if result.err != nil && cfg.IsTerminalError != nil && cfg.IsTerminalError(result.err) {
```

Comment:
Wouldn't that be better if `IsTerminalError` also handles `err != nil` part? Hence we can reduce one conditional `result.err != nil` from the if condition.

Reply:
Maybe this is just me, but as a consumer of `DoUntilQuorum`, I wouldn't expect `IsTerminalError` to be called with a `nil` error, as `nil` means "no error" and therefore it can't be a terminal error.

Reply:
This also means that implementations of `IsTerminalError` don't need to check for `nil` errors, which simplifies their implementation slightly.